service/storage_proxy.cc

```diff
@@ -2970,8 +2970,10 @@ storage_proxy::~storage_proxy() {
 }
 
 storage_proxy::storage_proxy(distributed<replica::database>& db, storage_proxy::config cfg, db::view::node_update_backlog& max_view_update_backlog,
-        scheduling_group_key stats_key, gms::feature_service& feat, const locator::shared_token_metadata& stm, locator::effective_replication_map_factory& erm_factory)
+        scheduling_group_key stats_key, gms::feature_service& feat, const locator::shared_token_metadata& stm, locator::effective_replication_map_factory& erm_factory,
+        distributed<service::storage_service>& ss)
```

Comment:
We are creating a (one more) circular dependency here. ss depends on qp and view_builder and both of them depend on the storage_proxy.

Reply:
What do you recommend? How to avoid the circular dependency?

Reply:
Why putting truncate code into ss in the first place? May be provide group0_client to storage_proxy::start_remote() and implement the truncation code directly there?

Reply:
Removing the truncate code from `storage_service`, and implementing it in `storage_proxy::remote` would require:
- Making a new `abort_source` (instead of the already existing one `storage_service::_group0_as`) which is needed in `raft_group0_client::start_operation()` and `raft_group0_client::add_entry()`
- Having access to `storage_service::_topology_state_machine` to re-implement the functionality in `storage_service::wait_for_topology_request_completion()`.

Maybe we could make a compromise where we pass `storage_service` to `storage_proxy::start_remote()`, and then we call `storage_service::truncate_table_for_tablets()` from a new `storage_proxy::send_truncate_for_tablet()`. That way we don't really have a circular reference. The only awkward thing would be having `storage_proxy::remote` "send" a topology request, instead of sending RPCs which it normally does.

Reply:
> Removing the truncate code from `storage_service`, and implementing it in `storage_proxy::remote` would require:
> 
>     * Making a new `abort_source` (instead of the already existing one `storage_service::_group0_as`) which is needed in `raft_group0_client::start_operation()` and `raft_group0_client::add_entry()`
> 

Not a huge problem. Can add a new abort source in storage_proxy::remote and trigger it in stop_remote.

>     * Having access to `storage_service::_topology_state_machine` to re-implement the functionality in `storage_service::wait_for_topology_request_completion()`.
> 

storage_service::_topology_state_machine is passed to storage_service from outside. It can be passed to start_remote as well.

Reply:
Yes, see 98323be296fdc96c6f63e15fa196c6de0ae349af, where repair service accesses topology_state_machine.

Reply:
The circular dependency has been resolved in v7